src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseApprovalService.cs

```diff
@@ -113,92 +113,155 @@ IEmailTemplateService emailTemplateService
+                    releaseId,
+                    releaseStatus.Id,
+                    request.PublishMethod == PublishMethod.Immediate);
+            }
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Unit>> SendEmailNotificationsAndInvites(
+            ReleaseStatusCreateRequest request,
+            Release release)
+        {
+            if (request.ApprovalStatus == ReleaseApprovalStatus.Approved
+                && request.PublishMethod == PublishMethod.Scheduled)
+            {
+                return await SendPreReleaseUserInviteEmails(release);
+            }
+
+            if (request.ApprovalStatus == ReleaseApprovalStatus.HigherLevelReview)
+            {
+                var userReleaseRoles =
+                    await _userReleaseRoleService.ListUserReleaseRolesByPublication(
+                        ReleaseRole.Approver,
+                        release.Publication.Id);
+
+                var userPublicationRoles = await _context
+                    .UserPublicationRoles
+                    .Include(upr => upr.User)
+                    .Where(upr =>
+                        upr.PublicationId == release.PublicationId
+                        && upr.Role == PublicationRole.Approver)
+                    .ToListAsync();
+
+                var notifyHigherReviewers = userReleaseRoles.Any() || userPublicationRoles.Any();
+
+                if (notifyHigherReviewers)
+                {
+                    return userReleaseRoles
+                        .Select(urr => urr.User.Email)
+                        .Concat(userPublicationRoles.Select(upr => upr.User.Email))
+                        .Distinct()
+                        .Select(email => _emailTemplateService.SendReleaseHigherReviewEmail(email, release))
+                        .OnSuccessAllReturnVoid();
+                }
+            }
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Unit>> SendPreReleaseUserInviteEmails(Release release)
+        {
+            var unsentUserReleaseInvites = await _context.UserReleaseInvites
+                .AsQueryable()
+                .Where(i =>
+                    i.ReleaseId == release.Id
+                    && i.Role == ReleaseRole.PrereleaseViewer
+                    && !i.EmailSent)
+                .ToListAsync();
+
+            var emailResults = await unsentUserReleaseInvites
+                .ToAsyncEnumerable()
+                .SelectAwait(async invite =>
+                {
+                    var user = await _userRepository.FindByEmail(invite.Email);
+                    return await _preReleaseUserService.SendPreReleaseInviteEmail(
+                            release,
+                            invite.Email.ToLower(),
+                            isNewUser: user == null)
+                        .OnSuccessDo(() => _preReleaseUserService.MarkInviteEmailAsSent(invite));
```

Comment:
If `invite` is passed to `SendPreReleaseInviteEmail` then it can be responsible for marking as sent making this logic more encapsulated, but I'm guessing you already attempted this change and then realised that not all usages of `SendPreReleaseInviteEmail` are marking as sent currently?

Reply:
Yeah exactly - I don't want to change any existing behaviour as part of this ticket.  Rather, I've suggested an investigation and implementation in ticket EES-4681.